r2-filter-compression/src/main/java/com/linkedin/r2/filter/compression/AbstractCompressor.java

```diff
@@ -0,0 +1,142 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+package com.linkedin.r2.filter.compression;
+
+import com.linkedin.data.ByteString;
+import com.linkedin.util.FastByteArrayOutputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import org.apache.commons.io.IOUtils;
+
+
+/**
+ * An abstract class housing common compression/decompression operations
+ */
+abstract class AbstractCompressor implements Compressor
+{
+  @Override
+  public byte[] inflate(InputStream data) throws CompressionException
+  {
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    InputStream inflaterStream = null;
+
+    try
+    {
+      inflaterStream = createInflaterInputStream(data);
+      IOUtils.copy(inflaterStream, out);
+    }
+    catch (IOException e)
+    {
+      throw new CompressionException(CompressionConstants.DECODING_ERROR + getContentEncodingName(), e);
+    }
+    finally
+    {
+      IOUtils.closeQuietly(inflaterStream);
+    }
+
+    return out.toByteArray();
+  }
+
+  @Override
+  public ByteString inflate(ByteString data) throws CompressionException
+  {
+    FastByteArrayOutputStream out = new FastByteArrayOutputStream();
+    InputStream inflaterStream = null;
+
+    try
+    {
+      inflaterStream = createInflaterInputStream(data.asInputStream());
+      IOUtils.copy(inflaterStream, out);
+    }
+    catch (IOException e)
+    {
+      throw new CompressionException(CompressionConstants.DECODING_ERROR + getContentEncodingName(), e);
+    }
+    finally
+    {
+      IOUtils.closeQuietly(inflaterStream);
+    }
+
+    return out.toUnsafeByteString();
+  }
+
+  @Override
+  public byte[] deflate(InputStream data) throws CompressionException
+  {
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    OutputStream deflaterStream = null;
+
+    try
+    {
+      deflaterStream = createDeflaterOutputStream(out);
+      IOUtils.copy(data, deflaterStream);
+    }
+    catch (IOException e)
+    {
+      throw new CompressionException(CompressionConstants.DECODING_ERROR + getContentEncodingName(), e);
+    }
+    finally
+    {
+      IOUtils.closeQuietly(deflaterStream);
+    }
+
+    return out.toByteArray();
+  }
+
+  @Override
+  public ByteString deflate(ByteString data) throws CompressionException
```

Comment:
minor: better put some javadoc comment to explain the advantage of this version than the other one.

Otherwise, this looks good to me.